Bluetooth: Mesh: Friendship improvements

Andrewpini opened the pull request:
- Adds Kconfig option to enable tuning the Friend receive delay window.
- Adds separate tag for friendship related messages
- Adds optional separate advertising set for friend related messages

LingaoM commented:
Pin：https://github.com/zephyrproject-rtos/zephyr/pull/52072#discussion_r1037325362

jhedberg commented on subsys/bluetooth/mesh/adv.h:
```diff
@@ -71,9 +72,10 @@ int bt_mesh_scan_disable(void);
 int bt_mesh_adv_enable(void);
 
 void bt_mesh_adv_buf_local_ready(void);
-
```
Any reason why you're removing this line, and thereby grouping `local` and `relay` together, but not `friend`?

Andrewpini replied:
No, just an oversight by me. Will fix 👍

LingaoM commented:
What I want to express is that if the PR about the priority relay I am doing now can be merged in the near future, we don't need to be so anxious to do a separate advertising set now, which will bring extra work.

LingaoM commented:
We currently support proxy occupying a separate advertising set, because it only has one persistent connection advertising, but for friends, the situation is different, there may be many friendships, using a separate advertising set can only alleviate the problem, it does not  It is not the best solution, so it may be a better choice to share multiple relay advertising sets.

PavelVPV commented:
> We currently support proxy occupying a separate advertising set, because it only has one persistent connection advertising, but for friends, the situation is different, there may be many friendships, using a separate advertising set can only alleviate the problem, it does not It is not the best solution, so it may be a better choice to share multiple relay advertising sets.

You are right that we could use an available relay set instead of the main to send friend message. But the problem is that a relay message can block friend message and it can block for quite long time (tens of ms, which is enough to miss the receive window). And a node relays messages more often and less predictable than sends messages between friend and lpns.

When using a separate set for friend, it usually takes ~2-5ms to schedule an advertisement and get back to mesh code. So the separate set can help to reduce the receive window and thus power consumption on LPN.

In case of multiple friendships, what is the probability that several lpns will poll a friend at the same time (with ~5ms error)? In this case the receive window can be increased to reduce the chance of missing it by friend, and I think this is fine since in most cases friend will still respond almost at that start of the receive window and lpn won't scan for full receive window.

The comment I posted in your PR is exactly to make friend more efficient even with just 1 adv set than it is right now. And yes, we can utilize the relay sets, this is a good optimization. But as I said, if you have already pushed a message to the controller, your friend message will be delayed. And this is less predictable than with the separate set.

PavelVPV commented:
> > When using a separate set for friend, it usually takes ~2-5ms to schedule an advertisement and get back to mesh code. So the separate set can help to reduce the receive window and thus power consumption on LPN.
> > In case of multiple friendships, what is the probability that several lpns will poll a friend at the same time (with ~5ms error)? In this case the receive window can be increased to reduce the chance of missing it by friend, and I think this is fine since in most cases friend will still respond almost at that start of the receive window and lpn won't scan for full receive window.
> 
> Yes, I'm note that in adv_ext.c has delays method, may this will decrease friendship efficient?
> 
> `k_work_reschedule(&adv->work, K_MSEC(ADV_INT_FAST_MS - delta));`

Yes, this needs to be addressed.

Andrewpini commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -423,6 +423,13 @@ config BT_MESH_ADV_EXT_GATT_SEPARATE
 	  Use a separate extended advertising set for GATT Server Advertising,
```
I agree that this could be improved. Could you create a separate PR with your suggested changes? I would prefer not to change it in this one.

mia-ko commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -875,6 +882,15 @@ config BT_MESH_FRIEND_SEG_RX
+	default 0
+	help
+	  Latency in milliseconds between request for and start of Friend
+	  advertising. Used to tune the receive delay window, making Friend
```
```suggestion
	  advertising. Used to tune the ReceiveDelay window, making Friend
```

mia-ko commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -421,7 +421,14 @@ config BT_MESH_ADV_EXT_GATT_SEPARATE
+	bool "Use a separate extended advertising set for Friend advertising"
+	depends on BT_MESH_FRIEND
+	help
+	  Use a separate extended advertising set for Friend advertising,
+	  otherwise a shared advertising set will be used.
```
...makes the Friend node...
+
ReceiveWindow?

mia-ko commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -875,6 +882,15 @@ config BT_MESH_FRIEND_SEG_RX
+	int "Latency for enabling advertising"
+	default 0
+	help
+	  Latency in milliseconds between request for and start of Friend
+	  advertising. Used to tune the receive delay window, making Friend
+	  poll events more efficient by compensating for the time the
+	  non-ideal latency otherwise would have added to the delay.
```
ReceiveDelay?
+
...thus compensating for the time between pushing the message to the Bluetooth host and the actual advertising start.

Andrewpini commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -333,6 +368,15 @@ void bt_mesh_adv_buf_relay_ready(void)
 
+void bt_mesh_adv_buf_friend_ready(void)
+{
+#if defined(CONFIG_BT_MESH_ADV_EXT_FRIEND_SEPARATE)
+	(void)schedule_send(&adv_friend);
+#else
+	(void)schedule_send(&adv_main);
+#endif
```
`adv_friend` is only available if CONFIG_BT_MESH_ADV_EXT_FRIEND_SEPARATE is defined, so the #ifdef must be present I guess.

Andrewpini committed:
Bluetooth: Mesh: Config opt for friend adv latency

Adds Kconfig option to enable tuning the Friend receive delay window.

Friend poll events can be made more efficient by compensating for the
time the non-ideal advertising latency otherwise would add to the receive
delay.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

Andrewpini committed:
Bluetooth: Mesh: Separate adv tag/set for friend

The purpose of this commit is to facilitate future improvements to the
LPN and friendship feature. By being able to identify friendship related
messages on the advertising layer, it will be possible to treat these in
a specific manner so that message exchange between LPN and friend
devices can be conducted in a more power efficient manner.

This commit adds the following:
   - A separate tag for friendship related messages
   - A optional separate advertising set for friend related messages

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

Andrewpini commented:
I've fixed the majority of the comments and refactored parts of `adv.c` in an attempt to make it a little more readable. @PavelVPV if you could espesially mind `bt_mesh_adv_buf_get_by_tag` an `bt_mesh_adv_send` to ensure that there is no misunderstanding on my part as of how it should work.

Andrewpini commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -143,6 +160,12 @@ struct net_buf *bt_mesh_adv_buf_get(k_timeout_t timeout)
+						K_POLL_MODE_NOTIFY_ONLY,
+						&bt_mesh_friend_queue,
+						0),
+#endif /* CONFIG_BT_MESH_ADV_EXT_FRIEND_SEPARATE */
```
Looks like your right 👍

PavelVPV commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -198,6 +219,9 @@ void bt_mesh_adv_buf_get_cancel(void)
 	k_fifo_cancel_wait(&bt_mesh_relay_queue);
 #endif /* CONFIG_BT_MESH_RELAY_ADV_SETS */
 
+#if CONFIG_BT_MESH_ADV_EXT_FRIEND_SEPARATE
```
`bt_mesh_friend_queue` is always compiled, so IS_ENABLED should work here.

PavelVPV commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -210,18 +234,24 @@ void bt_mesh_adv_send(struct net_buf *buf, const struct bt_mesh_send_cb *cb,
 	BT_MESH_ADV(buf)->cb_data = cb_data;
 	BT_MESH_ADV(buf)->busy = 1U;
 
+#if defined(CONFIG_BT_MESH_ADV_EXT_FRIEND_SEPARATE)
+	if (BT_MESH_ADV(buf)->tag == BT_MESH_FRIEND_ADV) {
```
`bt_mesh_friend_queue` and `bt_mesh_adv_buf_friend_ready` are always compiled, so IS_ENABLED should work here.

PavelVPV reviewed (dismissed):
Feel free to disregard the comments regarding `IS_ENABLED` as `IS_ENABLED` should work for relay queue as well, but now both relay and friend implementations are aligned with each other.

Last question. Did you do sanity check of friendship with `CONFIG_BT_MESH_ADV_EXT_FRIEND_SEPARATE` option enabled? Or did you try bsim friendship tests with the new option?

PavelVPV commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -159,22 +176,26 @@ struct net_buf *bt_mesh_adv_buf_get(k_timeout_t timeout)
+	}
+#endif
+
+	if (tag & BT_MESH_LOCAL_ADV || tag & BT_MESH_FRIEND_ADV || tag & BT_MESH_RELAY_ADV ||
```
On more small comment. Not sure if we need to check tag here. If we are not pooling friend neither relay message, then only locally created messages are left.

Andrewpini replied:
> Feel free to disregard the comments regarding `IS_ENABLED` as `IS_ENABLED` should work for relay queue as well, but now both relay and friend implementations are aligned with each other.
> 
> Last question. Did you do sanity check of friendship with `CONFIG_BT_MESH_ADV_EXT_FRIEND_SEPARATE` option enabled? Or did you try bsim friendship tests with the new option?

I ran the entire mesh bsim suite with every cfg combination I could think of, and it seems fine if we disregard the packet collision in friendship/ msg_group.sh that we have discussed earlier.

Andrewpini commented:
> Feel free to disregard the comments regarding `IS_ENABLED` as `IS_ENABLED` should work for relay queue as well, but now both relay and friend implementations are aligned with each other.
> 
> Last question. Did you do sanity check of friendship with `CONFIG_BT_MESH_ADV_EXT_FRIEND_SEPARATE` option enabled? Or did you try bsim friendship tests with the new option?

I ran the entire mesh bsim suite with every cfg combination I could think of, and it seems fine if we disregard the packet collision in friendship/msg_group.sh that we have discussed earlier.

alxelax reviewed (commented):
LGTM just a couple of comments

Andrewpini commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -159,22 +176,21 @@ struct net_buf *bt_mesh_adv_buf_get(k_timeout_t timeout)
+		return net_buf_get(&bt_mesh_friend_queue, timeout);
+	}
+
+#if CONFIG_BT_MESH_RELAY_ADV_SETS
```
(discussed in person)
As we saw, it seems there is not straight forward way to evaluate integer type configuration parameters like you can with boolean types in the `IS_ENABLED()` macro. While I agree that it would be better to align the style I really do prefer using `IS_ENABLED()` rather that `#ifdef` since I think it is more readable. Is this ok for you?

Andrewpini commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -878,6 +888,16 @@ config BT_MESH_FRIEND_SEG_RX
 	  many elements we can simultaneously be receiving segmented
 	  messages from when the messages are going into the Friend queue.
 
+config BT_MESH_FRIEND_ADV_LATENCY
```
According to Pavel's investigation it is >= 4ms. I'm not sure if this applies for all usecases. Tuning this value should be done by observation, trial and error but we could provide a range so the user will know the ballpark scope for this tuning.

Andrewpini commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -208,6 +223,21 @@ static int buf_send(struct bt_mesh_ext_adv *adv, struct net_buf *buf)
 	return err;
 }
 
+static const char *adv_tag_to_str(enum bt_mesh_adv_tag tag)
```
Sure, if you think that is necessary👍

PavelVPV replied:
Not sure if this is necessary. When logs are disabeld, the function is not used and will be compiled out.

alxelax replied:
If there is no function in the final elf then I'm Ok to leave it as is. @Andrewpini could you check?

Andrewpini replied:
I shall check 👍

alxelax commented:
@Andrewpini, seems PR needs rebase.
